src/compiler/context.js

```diff
@@ -10,219 +10,258 @@ const STATES = {
-        this.append(null, null, `      var on${indexHash} = td.${util.getTdMethodName('getNodeAtIdxPath')}(root, ${JSON.stringify(indexes)});
-      ${output};\n`);
-      }
+    incrementCurrentTdBody() {
+      tornadoBodiesPointer++;
     },
-
-    helper(node) {
-      let maxTdIndex = this.tdBodies.length - 1;
-      let tdIndex = tornadoBodiesPointer;
-      let indexes = this.htmlBodies[tdIndex].htmlBodiesIndexes;
-      let indexHash = indexes.join('');
-      let paramsHash, bodiesHash;
-      paramsHash = node.params.reduce((acc, param) => {
-        let paramVal = param.val;
-        if (Array.isArray(paramVal)) {
-          paramVal = `td.${util.getTdMethodName('get')}(c, ${JSON.stringify(paramVal[1].key)})`;
-        } else {
-          paramVal = `'${paramVal}'`;
-        }
-        acc.push(`${param.key}: ${paramVal}`);
-        return acc;
-      }, []);
-      paramsHash = `{${paramsHash.join(',')}}`;
-      bodiesHash = node.bodies.reduce((acc, body, idx) => {
-        let bodyName = body[1].name;
-        acc.push(`${bodyName}: this.r${maxTdIndex + idx + 2}.bind(this)`);
-        return acc;
-      }, []);
-      if (node.body.length) {
-        bodiesHash.push(`main: this.r${maxTdIndex + 1}.bind(this)`);
-      }
-      bodiesHash = `{${bodiesHash.join(',')}}`;
-      if (this.state !== STATES.HTML_ATTRIBUTE) {
-        this.append(null, `      ${util.createPlaceholder(this)};\n`,
-                          `      var on${indexHash} = td.${util.getTdMethodName('getNodeAtIdxPath')}(root, ${JSON.stringify(indexes)});
-        td.${util.getTdMethodName('helper')}('${node.key.join('')}', c, ${paramsHash}, ${bodiesHash}).then(function(val) {
-          td.${util.getTdMethodName('replaceNode')}(on${indexHash}, val);
-        });\n`);
-      }
+    decrementCurrentTdBody() {
+      tornadoBodiesPointer--;
     },
-
-    block(node) {
-      let tdIndex = tornadoBodiesPointer;
-      let indexes = this.htmlBodies[tdIndex].htmlBodiesIndexes;
-      let indexHash = indexes.join('');
-      if (this.state !== STATES.HTML_ATTRIBUTE) {
-        this.append(null, `      ${util.createPlaceholder(this)};\n`, null);
-        this.append(null, null, `      var on${indexHash} = td.${util.getTdMethodName('getNodeAtIdxPath')}(root, ${JSON.stringify(indexes)});
-      td.${util.getTdMethodName('replaceNode')}(on${indexHash}, td.${util.getTdMethodName('block')}('${node.blockName}', ${node.blockIndex}, c, this));\n`);
-      } else {
-        return `td.${util.getTdMethodName('nodeToString')}(td.${util.getTdMethodName('block')}('${node.blockName}', ${node.blockIndex}, c, this))`;
-      }
+    incrementHtmlEl() {
+      htmlElementPointer++;
     },
-
-    inlinePartial(node) {
-      console.log(node);
+    decrementHtmlEl() {
+      htmlElementPointer--;
     },
-    bodies(node) {
-      console.log(node);
+    pushInstruction(instruction) {
+      results.instructions.push(instruction);
+    },
+    stack: {
```

Comment:
this might be cleaner as a separate module. it'll at least be easier to unit test.